pkg/storegateway/batch_series_parallel.go

```diff
@@ -0,0 +1,137 @@
+package storegateway
+
+import (
+	"io"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb/chunks"
+	"github.com/thanos-io/thanos/pkg/pool"
+
+	"github.com/grafana/mimir/pkg/storegateway/indexheader"
+	"github.com/grafana/mimir/pkg/storegateway/storepb"
+)
+
+// batchSeriesSet will implement storepb.SeriesSet
+//
+// Implementation will be similar to existing batchedSeriesSet
+// A batchSeriesSet is created and used in BucketStore.Series
+type batchSeriesSet interface {
+	storage.SeriesSet
+
+	seriesWatcher
+}
+
+// batchSeriesSetFactory is used in BucketStore.Series
+type batchSeriesSetFactory func(
+	// postings can also become a postings provider, but for now we fetch all postings in one go.
+	// It can also be calculated by the seriesSet itself. For simplicity of this interface it's just provided.
+	postings []storage.SeriesRef,
+	batcher batcher,
+	batchFactory batchFactory, // the batchedSeriesSet creates its own batches and passes them to be populated
+) batchSeriesSet
+
+type seriesWatcher interface {
```

Comment:
Not sure about the name "watcher" cause this doesn't watch a series (or series set). This interface provides a method to release a series, so maybe `seriesReleaser` and then rename `seriesDisposed()` to `releaseSeries()`?